### ISSUE TITLE
question regarding input count matrix

### ISSUE THREAD

**Issue by bioarpit1** [YosefLab/MPRAnalyze#10]
Hi Thanks for making very useful statistical analysis tool MPRAnalyze, I am currently trying to use it for my data. I have a bit naive question regarding input matrix pattern for DNA and RNA count. currently my input matrix looks like this 
![image](https://user-images.githubusercontent.com/29548776/92064047-8522b680-ed51-11ea-91a2-fc3426dbca26.png)
                                                  rep1   rep2   rep3 
elementid_allele:barcode1         1         0         3 
elementid_allele:barcode2         11        20      23 
elementid_allele:barcode3         10         0         30 
elementn_allele:barcodeN          x          y         z 

i am bit confused regarding allelic input and how to analyze it 
currently after size estimation i used
 obj <- analyzeQuantification(obj = Th1, 
                              dnaDesign = ~ barcode)
this is running but i know there is something wrong here
[============>-------------------------------------------]  22% (71604/318937)
error fitting: rs3845302_G:AATGATGGAGC: Error in optim(par = guess, fn = cost.dnarna, llfnDNA = ll.funs$dna, llfnRNA = ll.funs$rna, : L-BFGS-B needs finite values of 'fn'

Basically in my experiment i am interested in allelic comparision of enhancers. 
1. how should my input look (where to add barcode level information, should i make another column?)
2. after estimating size what option should i use for analyzeQuantification

Thanks

Arpit

**Comment by talashuach** [comment]
Hi Arpit,

In terms of formatting the input, MPRAnalyze expects the input to be formatted as follows:
* two matrices: one for DNA counts, the other for RNA counts
* each sequence should be a single row in the matrix. For allelic comparisons, this means both alleles in the same row.
* an additional data.frame with column-annotations to provide information regarding each column.

In your case, this would look something like:

count matrix:
elementid       o1    o2    o3   o4  o5   o6         m1       m2       m3
rs10001246    14    11    17    1    11    2    ...    858      830      1237 ....
rs10012017    ....

And then in the annotation data.frame, you'll specify for each column which batch, barcode and allele they come from (note that barcode 1 from allele A is different from barcode 1 in allele B, so the annotations should reflect that.
In this case:

            batch    barcode    allele
o1        b1         A1             A
o2        b2         A1             A
o3        b3         A1             A
o4        b1         A2             A
o5        b2         A2             A
o6        b3         A2             A
....
m1       b1         B1             B
m2       b2         B1             B
m3       b3         B1             B
....

Note that the rownames for the annotations should match with the colnames of the count matrices.
Additionally - if the number of barcodes you have varies between elements, simply pad the missing values with 0s (MPRAnalyze will discard those observations internally).


That's for the input part of your question, I hope that's clear! Please let me know if it isn't.

For the downstream analyses, since you're actually comparing the two alleles, you should use analyzeComparative, since analyzeQuantification would only provide you with a quantification of the induced transcription rate, but doesn't compare anything.
Exactly how to call analyzeComparative depends on your specific study design (you can find more details in the package vignette).
If your data is paired (meaning, the DNA and RNA observations come from the same post-transfection library), then it should look something like:

`obj <- analyzeComparative(Th1, dnaDesign = ~ batch + barcode + allele, rnaDesign = ~ batch + allele, reducedDesign = ~ batch)`
After running this you'll have a comparative model for each sequence, which you can then test for significance using the `testLrt` function (it'll return p-values and logFC values for each comparison).

There are more details about the analyses in the package vignette, but I hope this was helpful.
I'll keep the issue open for now, in case you have any more problems running the package.

Good luck!
Tal

**Comment by bioarpit1** [comment]
Thanks for the response  

After reading your response i have two more question
1. "each sequence should be a single row in the matrix. For allelic comparisons, this means both alleles in the same row."
i have used custom index sequence for allignment where each allele is a different sequence thus i have allele wise count, for example 
rs1234_alleleA count 10, rs1234_alleleB count 20. So how can i put both alleles in a same row ? 

2. in the annotation file example 
o1 b1 A1 A
o2 b2 A1 A
o3 b3 A1 A
o4 b1 A2 A
i have some confusion so o1, o2...o4 is batch, b1.,b2,b3 is barcode, A or B is for allele so what is A1 , A2, here ? 

Sorry for my novice questions, and thanks for the prompt response, I appreciate your help.

**Comment by talashuach** [comment]
Happy to help!

1 - that's how allelic studies always work, it's not a problem. if you have N barcodes for allele A and M barcodes for allele B, your row would simply have (N+M) columns, with the observations for allele A followed by those for allele B. You'll tell the model which observation came from which allele using the annotation data.frame.
2 - Sorry, there was a formatting issue with my response. The first column (observation) is simply the rownames of the data.frame, and those should match with the colnames of the input data matrix (o1, o1, ... m1, m2... are the names I chose in this example for no reason whatsoever).
b1/b2/b3 are the batch indices, A/B are the allele identifier.
Then for the barcodes - since barcode #1 for allele A is a different barcode from barcode #1 for allele B, I simply give them different identifiers built from the barcode number + allele.

So the first row in the annotation, matching with the first column in the data, is replicate 1, allele A, barcode A1. Second column is replicate 2, allele A, barcode A1, then the fourth line is replicate 1, allele A, barcode A2 (= barcode #2 in allele A).

I hope this clarifies things!
I also fixed the formatting issue with the response above, just in case.

**Comment by bioarpit1** [comment]
Hi Tal, 

Thanks, I was able to use MPRAnalyze with my data but I have a few more queries regarding the result post LRT analysis
1. we are interested in allelic enhancer differences thus for creating reduce model i have selected allelic column
i have used these commands to build the model with allelic reduced 

objrs_allele <- analyzeComparative(objrs_depth, dnaDesign = ~ batch + barcode_allelic + allele, rnaDesign = ~ batch + allele, reducedDesign = ~ allele) 
Am i correct in assuming my significant enhancer hits are due to significant expression difference between alleles.

2. If my assumption that significant hits are due to the significant allelic difference based on my understanding of the model is it possible to get/plot alpha between allelic combination (something similar to the example given for alpha value plotting between the condition in quantitative analysis)

3. while using library size normalization when i use default uq normalization i get an error but using totsum is there any downside to it 

4. in my input for some enhancer only one allele is present (due to sequencing depth or post-transfection recovery) will that affect the model building specifically when looking at allelic differences 



Thank you very much and pardon me for my naive questions.  MPRAnalyze is a really very helpful tool I really appreciate your feedback.

**Comment by bioarpit1** [comment]
Thanks your email resolved this issue for reduced model i need to use batch if i want to see effect due to allele. Thanks for the tool and answers.
